types/coin_benchmark_test.go

```diff
@@ -70,3 +70,71 @@ func BenchmarkCoinsAdditionNoIntersect(b *testing.B) {
 		b.Run(fmt.Sprintf("sizes: A_%d, B_%d", sizeA, sizeB), benchmarkingFunc(sizeA, sizeB))
 	}
 }
+
+func BenchmarkSumOfCoinAdds(b *testing.B) {
+	// This benchmark tests the performance of adding a large number of coins
+	// into a single coin set.
+	// it does numAdds additions, each addition has (numIntersectingCoins) that contain denoms
+	// already in the sum, and (coinsPerAdd - numIntersectingCoins) that are new denoms.
+	benchmarkingFunc := func(numAdds, coinsPerAdd, numIntersectingCoins int, sumFn func([]Coins) Coins) func(b *testing.B) {
+		return func(b *testing.B) {
+			b.ReportAllocs()
+			addCoins := make([]Coins, numAdds)
+			nonIntersectingCoins := coinsPerAdd - numIntersectingCoins
+
+			for i := 0; i < numAdds; i++ {
+				intersectCoins := make([]Coin, numIntersectingCoins)
+				num := NewInt(int64(i))
+				for j := 0; j < numIntersectingCoins; j++ {
+					intersectCoins[j] = NewCoin(coinName(j+1_000_000_000), num)
+				}
+				addCoins[i] = intersectCoins
+				for j := 0; j < nonIntersectingCoins; j++ {
+					addCoins[i] = addCoins[i].Add(NewCoin(coinName(i*nonIntersectingCoins+j), num))
+				}
+			}
+
+			b.ResetTimer()
+
+			for i := 0; i < b.N; i++ {
+				sumFn(addCoins)
+			}
+		}
+	}
```

Comment:
In contrast to declaring a type inside a function as I mentioned in another PR, declaring a function inside another function has subtleties worth avoiding:

- it's not always immediately obvious to the reader whether the function captures a closure around any other variables -- in this case, no
- stack traces are less clear as the anonymous function will have a numeric suffix of the parent function

I would prefer this written as a package-level function, perhaps named `benchmarkSumCoins` with signature `benchmarkSumCoins(b *testing.B, numCoins int, ...)` -- then on L135-137, the `b.Run` call looks more like:

```go
b.Run(fmt.Sprintf(...), func(b *testing.B) {
  benchmarkSumCoins(b, numAdds, ...)
})
```

Ultimately I think that's simpler to read and follow than an anonymous function that itself returns a closure.

Reply:
This is currently the standard written in other benchmarks in this package. Can this be done as an issue to refactor them all to this desired approach?

Reply:
I wasn't aware it was that way everywhere, feel free to leave it as-is for consistency. I have an offline list of cleanups that need to get turned into issues, and I'll add this one too.